src/engine/bufferscalers/enginebufferscalerubberband.cpp

```diff
@@ -286,3 +281,40 @@ int EngineBufferScaleRubberBand::runningEngineVersion() {
     return 2;
 #endif
 }
+
+void EngineBufferScaleRubberBand::reset() {
+    m_pRubberBand->reset();
+
+    // As mentioned in the docs (https://breakfastquay.com/rubberband/code-doc/)
+    // and FAQ (https://breakfastquay.com/rubberband/integration.html#faqs), you
+    // need to run some silent samples through the time stretching engine first
+    // before using it. Otherwise it will eat add a short fade-in, destroying
+    // the initial transient.
+#if RUBBERBANDV3
+    size_t remaining_padding = m_pRubberBand->getPreferredStartPad();
+#else
+    // This _should_ be equal to the latency in older Rubber Band versions:
+    // https://github.com/breakfastquay/rubberband/blob/c5f99d5ff2cba2f4f1def6c38c7843bbb9ac7a78/main/main.cpp#L652
+    size_t remaining_padding = m_pRubberBand->getLatency();
+#endif
+    std::fill_n(m_buffers[0].span().begin(), kRubberBandBlockSize, 0.0f);
+    std::fill_n(m_buffers[1].span().begin(), kRubberBandBlockSize, 0.0f);
+    while (remaining_padding > 0) {
+        const size_t pad_samples = std::min<size_t>(remaining_padding, kRubberBandBlockSize);
+        m_pRubberBand->process(m_bufferPtrs.data(), pad_samples, false);
+
+        remaining_padding -= pad_samples;
+    }
+
+#if RUBBERBANDV3
+    size_t padding_to_drop = m_pRubberBand->getStartDelay();
+#else
+    size_t padding_to_drop = m_pRubberBand->getLatency();
+#endif
+    while (padding_to_drop > 0) {
```

Comment:
> I was also also concerned how big the internal buffer of rubberband is, since I see here memory allocation:
https://github.com/breakfastquay/rubberband/blob/9001b5459c1c424a71be0829c8988a25a9edf490/src/faster/StretcherProcess.cpp#L487
Which will likely block our real time thread.

The padding is half of the window size, so that should definitely fit in Rubber Band's buffers. And thanks to the `m_pRubberBand->setTimeRatio(2.0)` call the buffer should be able to fit at least two of those windows before reallocating.

> In case of Mixxx, and after seeking within a track, it is not true that these samples are zero. So I think it is required to use the real samples before the start instead of zero.

That was part of my experiments in solving #11125, but it's a bit more difficult than it seems. Especially because the read ahead manager is shared by multiple things and reading data through it invokes a bunch of side effects. Older samples also may not be cached (I've noticed this causing issues when beat jumping somewhere for the first time that weren't present when not reading past samples). I have not yet had time to dig into this further, so if you have suggestions for how to compensate for the latency here without messing with the rest of the deck's state (so switching back and forth between linear interpolation and Rubber Band time stretching should be seamless and not shift any timings around) then that would be great.

> The confusing point is that we also have the available() function. In theory there must not bee any output available before not receive more than a whole m_aWindowSize. Is this the case?

Yes, like I mentioned before, they're using an STFT (Short-time Fourier Transform). It's an overlap-add process where you a window function to a chunk of the input data, take the discrete Fourier transform of that, process the frequency domain signal, take the inverse discrete Fourier transform of that, apply another window function, and then add that to the output buffer, partially overlapping the previous output. The reason why the required padding is half of the input buffer is because most window functions used for these things are raised cosine windows, like the Hann/Hanning function I linked below. Because you apply (= multiple the chunk of audio with) the window twice, once before the DFT, and once after the IDFT, you've effectively applied the square of the window function to the signal if you didn't do anything to the signal in the frequency domain. And if you try adding up squared Hann functions with a quarter period difference between them, you'll notice that after a period the resulting function forms a flat line. That explains why the required padding is half a window, and why you get that smooth fade in if you don't add the padding.

Some wikipedia inks if you're curious about how and why this all works:

https://en.wikipedia.org/wiki/Short-time_Fourier_transform
https://en.wikipedia.org/wiki/Overlap%E2%80%93add_method
https://en.wikipedia.org/wiki/Hann_function

Reply:
But yeah I guess you are right, we shouldn't be reading and dropping samples immediately. We should be doing the same thing the rubberband cli is doing, and process the audio as normal but drop the first `getStartDelay()` samples of it. That also aligns with the STFT primer I posted above heh.